### ISSUE TITLE
Session timed out due to inactivity

### ISSUE THREAD

**Issue by maarguello** [osTicket/osTicket#6335]
I'm experimenting issues login, osticket is showing a **msg** "Session timed out due to inactivity". it's just happening when use Chrome.

osTicket Version v1.17 
Web Server Software Apache/2.4.53 (Win64) OpenSSL/1.1.1n PHP/8.0.18
Chrome Versión 98.0.4758.102
Windows 11.

👋🏾

**Comment by JediKev** [comment]
@maarguello 

That means the session timed out as there was no activity to refresh it before it expired. For now you can try setting your Agent Session Timeout in your system settings to `0` which should make the session expiration 24 hours. We have a massive session pull being worked on and reviewed at the moment (not complete yet). Hopefully that will resolve your issue. Once the pull is approved it will be included in the next release.

Cheers.

**Comment by maarguello** [comment]
@JediKev

 understandood!, thanks for your quick response and the advice.

**Comment by darrell-badenhorst** [comment]
> @maarguello
> 
> That means the session timed out as there was no activity to refresh it before it expired. For now you can try setting your Agent Session Timeout in your system settings to `0` which should make the session expiration 24 hours. We have a massive session pull being worked on and reviewed at the moment (not complete yet). Hopefully that will resolve your issue. Once the pull is approved it will be included in the next release.
> 
> Cheers.

Good day

May I ask if there has been any further development with this condition as yet?

Regards

**Comment by JediKev** [comment]
@darrell-badenhorst 

If you are using v1.17.2 with the following patches applied you shouldn't experience further session issues:
- https://github.com/osTicket/osTicket/pull/6379
- https://github.com/osTicket/osTicket/pull/6434

After applying these changes clear the cache, cookies, and sessions and retest. Please let us know if you run into this again (I cannot replicate this so I'll need exact steps if it happens again).

Cheers.

**Comment by darrell-badenhorst** [comment]
Hi there

thank you for the reply, much appreciated. We are on osTicket (v1.16.5). Will upgrade and test.

**Comment by mrkylegp** [comment]
@JediKev I have applied the patches above and "agent session timeout" is set to 0 (to disable) yet as an agent I still get logged out after a while, regardless of having cleared cache and accessing osticket on different computers. Using Chrome. Is there something I need to clear on the server side and not just the client side?

**Comment by JediKev** [comment]
@mrkylegp 

If you are running PHP-FPM make sure you restart it.

Cheers.

**Comment by mrkylegp** [comment]
> @mrkylegp
> 
> If you are running PHP-FPM make sure you restart it.
> 
> Cheers.

@JediKev 
Done this, even rebooted the whole server. Still times out and shows the "Authentication Required" screen.

**Comment by JediKev** [comment]
@mrkylegp 

Are you running v1.17.2? Have you tried clearing all cache, cookies, and sessions? Maybe try an incognito/private window as well.

Cheers.

**Comment by mrkylegp** [comment]
> @mrkylegp
> 
> Are you running v1.17.2? Have you tried clearing all cache, cookies, and sessions? Maybe try an incognito/private window as well.
> 
> Cheers.

@JediKev 
Yes, V1.17.2, I have tried all of those and regardless it still seems to occur.

I log in, work for a while, go back to the tab and it's showing "Authentication Required" again most of the time.

**Comment by JediKev** [comment]
@mrkylegp 

Have you tried increasing your Agent Session Timeout setting (**Admin Panel > Settings > Agents > Agent Session Timeout**)? Please note, if you set this to `0` the session should timeout after 24 hours.

Cheers.

**Comment by mrkylegp** [comment]
> @mrkylegp
> 
> Have you tried increasing your Agent Session Timeout setting (**Admin Panel > Settings > Agents > Agent Session Timeout**)? Please note, if you set this to `0` the session should timeout after 24 hours.
> 
> Cheers.

@JediKev It's currently set to 0, yet the sessions seem to time out anywhere between 20-60 minutes. Same browser too, same PC. Though I've tried on other browsers and PCs which have been completely cleared cache and cookie wise yet this still seems to happen. I don't believe I had this issue in v1.16.6. Since I have it set to 0 (to disable), it never timed out the session.

**Comment by mrkylegp** [comment]
Still occurring in v1.18.1. Even though session timeout is 0, it still gives me this problem.

**Comment by barkermn01** [comment]
I Will confirm I'm also having this problem, Logged in as an Admin user go to the home after a couple of seconds (so it tried to auto-refresh the Ticket List, if i try to visit any page i get "Session Timed out due to inactivity" and even if i re-logging it gives the same error "Session timed out due to inactivity", i have to use the browsers inspector to kill the OSTSESSID cookie to be allowed to log back in.

If you beat the list of Tickets in the SCP auto refreshing you don't have this problem it is something to do with the SCP Tickets list Auto-refreshing that breaks it.

Edit: So just watched it with the browser console request to `/scp/ajax.php/queue/counts`  triggers response of below and that is what then triggers the session timed out because it's not refreshing the token because of the missing CSRF Token for the AJAX request

```
403 Forbidden:

Must login: Valid CSRF Token Required
```

**Comment by leons1767** [comment]
I downloaded 1.17.6 and did a fresh installed. It seems the problem still pretty much there. Does anyone have a work around solution? Adjusting the session timeout setting to 0 doesn't work for me. Appreciate any tips.

**Comment by maarguello** [comment]
Hi @JediKev after upgrade to 1.17.6, the issue persist.

<img width="196" alt="Image" src="https://github.com/user-attachments/assets/da74a13e-45d4-45b2-ba78-8d32854e5f70" />

**Comment by casper-code80** [comment]
same issue here, has someone fixed it?

**Comment by JediKev** [comment]
If y'all [revert this commit](https://github.com/osTicket/osTicket/pull/6766/commits/416b548bd812188fec62c89fe8bfb0a3128731a1) does it start working again?

**Comment by casper-code80** [comment]
thanks for response. it did not helped in our case. Our main problem is that after timeout browser cookies needs to be cleared in order to login again.

**Comment by JediKev** [comment]
@casper-code80 

Okay [what about reverting this](https://github.com/osTicket/osTicket/pull/6796/files)? I believe this annd the previously link commit are the only two changes to session related code.

Cheers.

**Comment by casper-code80** [comment]
Unfortunately, no luck. I performed a fresh local install and reverted these two commits, but the problem persists. If I log out manually, I can log in again without any issues. I noticed that after a session timeout, the session cookie value does not change, whereas when logging out manually, it is renewed.

**Comment by JediKev** [comment]
@casper-code80 

Well I'm not too sure then. I cannot replicate this in our cloud environments; not even locally. Plus, the above two commits are the only two commits having anything to do with sessions. Are you sure you are using a supported PHP version (8.2.x - 8.4.x)?

Since you seem to be able to replicate the issue consistently you can help narrow this down by doing multiple installs/tests with v1.18.1 and reapplying the latest commits from v1.18.2 in small batches until you can identify which one causes the issue.

Cheers.

**Comment by casper-code80** [comment]
We are currently using version 1.17.6. We tested an upgrade to 1.18.2, but the issue was present there as well. After performing fresh installs of a few older versions, it appears that the issue was introduced in v1.16.4. Reverting [this commit](https://github.com/osTicket/osTicket/commit/1630459a21dae0124913ecfad20f2f59695453a2) seems to resolve the issue.

**Comment by liverpoolfc-fan** [comment]
I just ran into this yesterday too. Latest 1.18.2.  Interestingly, even though I have set session timeout to 30 minutes in the console years ago, following the advice above that clearing the cookie allows login I went to look for it. The cookie shows an expiry of 12:34 today. That would be a full 24 hours after I logged in yesterday. Seems something is not right with the setting of the expiry.

**Comment by liverpoolfc-fan** [comment]
A follow-on from my last post. 

When I login I am presented with the Ticket List. The cookie expiry is set at 24 hours (time off by 1 hour but guessing that is Daylight Savings). However, as soon as I click into a ticket, or switch to another tab, the cookie expiry updates to the duration stored in the Admin Console (30 mins in my case). So, it seems that it is just the initial login that is ignoring the defined timeout. But, if you are in this stuck state you can never click into a ticket or change tabs so the timeout never gets shortened.

Also, if it might help to narrow down the issue: We use the LDAP/AD authentication plug-in. When I first encountered the issue one of our two Domain Controllers was in the process of loading patches so may not have been available to authenticate the user. Once the user account was in thee stuck state it didn't matter that the Domain Controller came back online. An InCognito session could login but the original session kept getting the problem. The only cure (beyond waiting the 24 hours incorrectly set on the cookie) was to delete the cookie.

**Comment by its-vincent** [comment]
@JediKev I ran into this issue again today and had a little time available to debug, 

When the user enters their userid/password - the code in /scp/login.php 
`        $redirect($user->isValid() ? $dest : 'login.php');
`
throws them back to the login.php page because isValid() returns false. 

This is from the function 
class.usersession.php -> isvalidSession 
at the lines 
`        if ($maxidletime && ((time()-$expire) > $maxidletime))
            return false;
`
So, even though the user has just logged in and should be starting a new session - the old still in place session token is being used to compare the expiry time. 

Can you please tell me if this would be a vaild fix for the issue - or if it could introduce any other problems. 

In /scp/login.php 

If the $user object gets returned we have a valid logged in user I believe, so call refreshSession before redirecting them. 

```

    //if ($user = StaffAuthenticationBackend::process($username,
    //        substr($_POST['passwd'], 0, 128), $errors)) {
    //    $redirect($user->isValid() ? $dest : 'login.php');
    //}
    if ($user = StaffAuthenticationBackend::process($username,
            substr($_POST['passwd'], 0, 128), $errors)) {
        $user->refreshSession();
        $redirect($user->isValid() ? $dest : 'login.php');
    }


```
Please feel free to hide/delete this post if this would not be valid. 

Thanks

**Comment by burumart** [comment]
We have the same issue, we log in and te next day we have the error "Session timed out due to inactivity" and we can't login again until we clear all sessions.

**Comment by JediKev** [comment]
@burumart 

Try reverting [this patch](https://github.com/osTicket/osTicket/pull/6766/commits/416b548bd812188fec62c89fe8bfb0a3128731a1), clear all sessions/cookies in the browser, do a fresh login, and retest for a day or so.

Cheers.

**Comment by its-vincent** [comment]
@JediKev 

I don't think that reverting that patch will help. The problem is not that sessions are timing out too quickly. The problem is that once the user loads the login page the system loads the last stored cookie. This cookie contains whatever was the correct expiry the last time the user logged in. When the user attempts to login, the system examines this cookie to see if it the current session is valid. Because the cookie contains an expired timestamp it rejects the session as invalid and throws the user back to the login page. 

I believe the scenario is:- 
User logs in, does their work, maybe stays in a ticket
User closes browser without logging out
User reopens browser some time after the session cookie has expired, and the tab is restored
User attempts to interact with the system
System detects that the session has expired and throws them back to the login page to re-login 
Login page looks at the last stored cookie and determines that the current session has expired so it throws them back to the login page 
Endless loop! 

My suggested fix above is to prevent the endless loop. If the user successfully authenticates, refresh the session cookie before continuing.

**Comment by JediKev** [comment]
@its-vincent 

I get it and I think I have a hunch why. In any case, we'll have to revisit this later once we have time.

Cheers.

**Comment by its-vincent** [comment]
@JediKev 
I appreciate you guys are busy working on the upgrade.  

I just wanted to update you that I have been running with this change [issuecomment-3303256336](https://github.com/osTicket/osTicket/issues/6335#issuecomment-3303256336) in place for the last week and have not seen any obvious side-effects. I have deliberately left sessions open overnight/keeping tabs open/closing the browser and re-opening it the next day/etc. I get asked to re-login due to an expired session as expected, but no more endless loops.

**Comment by dixonmartinez** [comment]
@its-vincent https://github.com/osTicket/osTicket/issues/6335#issuecomment-3303256336

Was this the only modification you made?

It doesn't work in my environment.

I had to modify the following: In the class.ostsession.php file, I had to comment out this line:

https://github.com/osTicket/osTicket/blob/53339df0f5c5b0fcfc109600d62646a19c1db7e7/include/class.ostsession.php#L420

And in the class.usersession.php file, I added this to the line:

```php
// Is it expired?

if ($maxidletime && ((time()-$expire) > $maxidletime)){ 
session_unset(); 
osTicketSession::destroyCookie(); 
session_destroy(); 
return false; 
}
```

The method isvalidSession($htoken, $maxidletime=0, $checkip=false)

it has remained like this

```php
function isvalidSession($htoken, $maxidletime=0, $checkip=false){
        global $cfg;

        // Compare session ids
        if (strcmp($this->getSessionId(), session_id()))
            return false;

        $token = rawurldecode($htoken);
        // Check if we got what we expected....
        if ($token && !strstr($token,":"))
            return false;

        // Get the goodies
        list($hash, $expire, $ip) = explode(':', $token);

        // Make sure the session hash is valid
        if ((md5($expire . SESSION_SECRET . $this->userID) != $hash))
            return false;

        // is it expired??
        if ($maxidletime && ((time()-$expire) > $maxidletime)){
            session_unset();
            osTicketSession::destroyCookie();
            session_destroy();
            return false;
        }

        // Make sure IP is still same - if requested
        if ($checkip && strcmp($ip, MD5($this->getIP())))
            return false;

        $this->validated = true;

        return true;
   }
```

@JediKev 
I don't fully understand the methodology used for session management, or the system flow. If you could provide more information or point me to documentation explaining the osTicket process flow, it would be a great help in working on this solution. I'm not sure if it's the correct one, but it's the one that has worked for me.

**Comment by its-vincent** [comment]
@dixonmartinez Yes. I only changed that one line. 

Our environment - in case it makes any difference. 

Rocky 9, with php 8.2.28-1 from appstream, and mariaDB 10.5.29-2 from appstream
osTicket Core, v1.18.2
plugins audit.phar and auth-ldap.phar

**Comment by dixonmartinez** [comment]
When accessing the osTicket system to log in as an agent, I can see that a cookie and a database session are created.

<img width="1599" height="747" alt="Image" src="https://github.com/user-attachments/assets/9c973f71-8ba4-4791-87ba-8d5f38156f81" />

<img width="1285" height="493" alt="Image" src="https://github.com/user-attachments/assets/bde8e023-6532-4a10-8a04-f16b735b9b57" />

After entering the username and password and logging in:

<img width="1598" height="558" alt="Image" src="https://github.com/user-attachments/assets/36f9a2d7-11d9-4fa9-85e2-af2e99482bbc" />

If the login is successful, a new session is created

<img width="1257" height="531" alt="Image" src="https://github.com/user-attachments/assets/dacc56ae-54b3-4ff3-abd1-3038a9066f2e" />

And at the same time another cookie record is generated

<img width="1598" height="739" alt="Image" src="https://github.com/user-attachments/assets/d44b06af-be70-4232-a23f-ab5f36e995f0" />

I don't know if This must be correct.

Here I share information about my development environment:

<img width="967" height="694" alt="Image" src="https://github.com/user-attachments/assets/7a5f8433-6e03-46fb-9b3c-0e58b7ec44b2" />

<img width="1009" height="424" alt="Image" src="https://github.com/user-attachments/assets/5739dc90-7ec7-4d65-bc54-2bd2e370006b" />

I forked the osTicket project, and I had to modify the Schema Signature to be able to import the database.

```sql

SELECT session_id, session_data, session_expire, NOW() as db_now, now() > session_expire  IsExpired
FROM ost_session 
WHERE user_ip = '::1'
ORDER BY SESSION_Expire
;
```

**Comment by its-vincent** [comment]
I note you downloaded/forked the code from the git repository. I did not. I downloaded the released packages from the website. 

It is possible you have some changes from the git that were not in the release package. 

With the released packages, I just needed to make the one line change to get rid of the session timeout problem.

**Comment by ramalhoruben** [comment]
Hi all,

@JediKev I can confirm this problem remains unfixed as of v1.17.7 for Staff Session, and it also very likelly applies to ClientSession too, i think i found what is leading us to this problem, but i think if this is it, it would be very strange for dev team not to be able to replicate it and that it went unnoticed this long, but it is indeed a subtle detail..

Here's what I found:
 **class.usersession.php** and ClientSession and StaffSession populate **$staff->maxidletime** from system configuration.

`class ClientSession extends EndUser {
    use UserSessionTrait;

    function __construct($user) {
        global $cfg;
        parent::__construct($user);
        $this->class ='client';
        // XXX: Change the key to user-id
        $this->session = new UserSession($user->getUserId());
        $this->setSessionToken();
        $this->maxidletime = $cfg->getClientTimeout();
    }

    function getSessionUser() {
        return $this->user;
    }

    function isValid() {
        return $this->isValidSession();
    }
}

class StaffSession extends Staff {
    use UserSessionTrait;

    static function lookup($var) {
        global $cfg;
        if (($staff = parent::lookup($var))) {
            $staff->class = 'staff';
            $staff->session = new UserSession($staff->getId());
            $staff->setSessionToken();
            $staff->maxidletime = $cfg->getStaffTimeout();
            $staff->checkip = $cfg->enableStaffIPBinding();
        }
        return $staff;
    }

    function clear2FA() {
        unset($_SESSION['_auth']['staff']['2fa']);
        $_SESSION['_auth']['staff']['2fa'] = null;
        return true;
    }

    // If 2fa is set then it means it's pending
    function is2FAPending() {
        return isset($_SESSION['_auth']['staff']['2fa']);
    }

    function isValid() {
        return (!$this->is2FAPending() && $this->isValidSession());
    }
}`

As pointed out by @its-vincent , **scp/login.php** will authenticate user, check if session is valid and if not valid redirect user back to same page 'login.php', leading to a re-authentication loop until either user clear session data (cookies) from its browser or the condition that leads to this bug change, often through additional waiting time.
It so happens that isValidSession in **include/class.usersession.php** will check for session expiration like this **($maxidletime && ((time()-$expire) > $maxidletime))** both time and expire are measured in seconds, but maxidletime is set in minutes, thus leading to this particular problem, setting staff session timeout to zero in config solves this problem ofc, as this check is ignored completly.  

I just proposted a fix for this problem #6920

**Comment by JensEB** [comment]
With v1.18.3 I run into the same issue.
In my test system I can login, in my live system I can't.
I found the problem in my case.

The error is in class.ostsession.php by DatabaseSessionRecord::lookupRecord()
```
        catch (DoesNotExist $e) {
            // We're auto-creating model (unsaved) when one doesn't exist?
            $record = ($autocreate && ctype_alnum($id)) ? self::create($id) : null;
        }
```

This part checks the $id with ctype_alnum().
This is no good idea, because this may break valid sessions.

What do you set in your php settings for session.sid_bits_per_character?
If you set it to 4(default) or 5, it may work. But if you set it to 6, it breaks, because the session id can contain comma and minus.
https://www.php.net/manual/en/session.configuration.php#ini.session.sid-bits-per-character

change
` $record = ($autocreate && ctype_alnum($id)) ? self::create($id) : null;`
into
` $record = $autocreate ? self::create($id) : null;`
and it will work again.

If you want to check the ID, you can use following line:
` $record = ($autocreate && preg_match('/^[A-Za-z0-9,-]+$/', $id)) ? self::create($id) : null;`

This works for me, too.

**Comment by ramalhoruben** [comment]
@JensEB Mine is set to 5. That is the point with OsTicket, installation is quite subtle, and an improper installation might break things and leave us scratching our heads.. had one such problem when I set my osticket image TZ to Europe/Lisbon instead of UTC, it broke things subtly but enough for me to lose an afternoon figuring why drafts weren't saving. 
But i guess that is exactly the point of EnhanceSoft, they make a living  out of their hosting and cloud infrastructure, from good installs, and stable platforms, not from the software itself, which is gladly is open source.

**Comment by JensEB** [comment]
Yes, you're right. Every server has a different configuration, and it's difficult to keep track of everything. I'm also glad that osTicket is open source. 
This allows others to participate in further development...

**Comment by tsatrom** [comment]
We hit this in production (v1.18.2, Ubuntu 24.04, Apache mod_php, agents signing in with Google via the OAuth2 plugin) and found the root cause. Posting because it explains why this is so hard to reproduce and why clearing the cookie is the only escape.

**Root cause**

`UserSessionTrait::setSessionToken()` in `include/class.usersession.php` keeps a pre-existing token when called with no argument:

```php
function setSessionToken($token=null) {
    $this->token = &$_SESSION[':token'][$this->class];
    $token = $token ?: $this->token;   // keeps the OLD token if the session already has one
    $this->token = $token ?: $this->getSessionToken();
}
```

`StaffAuthenticationBackend::login()` calls `$staff->setSessionToken();` with no argument. So when an agent re-authenticates inside a session that still holds an expired token (idle past the agent timeout, browser never explicitly logged out, so `signOut()` never unset `$_SESSION[':token']['staff']`), the login itself succeeds but the stale token survives. The next request then fails `isvalidSession()` on the `(time()-$expire) > $maxidletime` check and bounces back to the login page with "Session timed out due to inactivity". Every retry repeats this, because every successful login keeps the same stale token. Deleting the OSTSESSID cookie works because a fresh empty session finally gets a fresh token.

**Why it is hard to reproduce**

You need all three at once: (1) idle past the agent timeout so the token timestamp goes stale, (2) no explicit logout (logout clears the token), and (3) a re-login attempt in that same browser session. Testing with a fresh browser profile, or after a logout, will never show it.

With SSO backends it looks even stranger: the OAuth2 callback (`/api/auth/oauth2`) completes the sign-in and redirects to the destination page, the audit log records a successful login, and the destination page immediately bounces back to the login screen. I believe #6222 is this same bug seen from that angle.

**Fix that worked for us**

One line in `StaffAuthenticationBackend::login()` (line 664 in v1.18.2), forcing a fresh token on every successful login:

```php
// Set session token
$staff->setSessionToken($staff->getSessionToken());
```

Applied today, immediately unstuck a session that had been looping for hours, with no cookie clearing. `UserAuthenticationBackend::login()` has the identical `$user->setSessionToken();` so the client portal presumably has the same loop; we applied the same one-line change there.

This also lines up with @its-vincent's findings above: calling `refreshSession()` after login fixes the password path for the same reason (the deadband sees the stale timestamp and renews the token), but making the token fresh inside `login()` also covers external/SSO backends, which never pass through `scp/login.php`.
